File: Codeforces/288D-Tanya-and-Password.py
hossz = int(input())
l = {}

for _ in range(hossz):
    uj = input()
    fst = uj[:2]
    snd = uj[1:]

    if fst in l:
        l[fst][1].append(snd)
    else:
        l[fst] = (0, [snd])

    if snd in l:
        l[snd] = (l[snd][0] + 1, l[snd][1])
    else:
        l[snd] = (1, [])


# https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/
# Python3 function to print Eulerian circuit in given
# directed graph using Hierholzer algorithm

def printCircuit(adj):

        # adj represents the adjacency list of
        # the directed graph
        # edge_count represents the number of edges
        # emerging from a vertex
    edge_count = dict()

    for i in adj:

        # find the count of edges to keep track
        # of unused edges
        edge_count[i] = len(adj[i])

    if len(adj) == 0:
        return  # empty graph

    # Maintain a stack to keep vertices
    curr_path = []

    # vector to store final circuit
    circuit = []

    # start from any vertex
    curr_path.append(curr)
    curr_v = curr  # Current vertex

    while curr_path:

        # If there's remaining edge
        if edge_count[curr_v]:

            # Push the vertex
            curr_path.append(curr_v)

            # Find the next vertex using an edge
            next_v = adj[curr_v][-1]

            # and remove that edge
            edge_count[curr_v] -= 1
            adj[curr_v].pop()

            # Move to next vertex
            curr_v = next_v

        # back-track to find remaining circuit
        else:
            circuit.append(curr_v)

            # Back-tracking
            curr_v = curr_path[-1]
            curr_path.pop()

    # we've got the circuit, now print it in reverse
    ordered[:] = circuit[:0:-1]


ends = [(l[a][0] - len(l[a][1]), a) for a in l if l[a][0] != len(l[a][1])]

empty = not ends
if empty:
    curr = next(iter(l.keys()))
    ends.append((-1, curr))
    ends.append((1, l[curr][1][0]))
else:
    ends.sort()

if len(ends) != 2 or ends[0][0] != -1 or ends[1][0] != 1:
    print("NO")
    exit()

curr = ends[0][1]
l[ends[1][1]][1].append(ends[0][1])

ordered = []
printCircuit({x: l[x][1] for x in l})

if not empty:
    for i in range(len(ordered)-1):
        if ordered[i] == ends[1][1] and ordered[i+1] == ends[0][1]:
            ordered = ordered[i+1:] + ordered[:i+1]
            break

if len(ordered) != hossz+1:
    print("NO")
else:
    print("YES")
    print(''.join(ordered)[::2] + ordered[-1][-1])

File: Codeforces/288D-Tanya-and-Password.cpp

// https://www.geeksforgeeks.org/hierholzers-algorithm-directed-graph/
// A C++ program to print Eulerian circuit in given
// directed graph using Hierholzer algorithm
#include <bits/stdc++.h>
using namespace std;

vector<int> printCircuit(vector<vector<int>> l)
{
    // l represents the adjacency list of
    // the directed graph
    // edge_count represents the number of edges
    // emerging from a vertex
    vector<int> edge_count;

    for (int i = 0; i < l.size(); i++)
    {
        //find the count of edges to keep track
        //of unused edges
        edge_count.push_back(l[i].size());
    }

    // if (!adj.size())
    //     return; //empty graph

    // Maintain a stack to keep vertices
    stack<int> curr_path;

    // vector to store final circuit
    vector<int> circuit;

    // start from any vertex
    curr_path.push(0);
    int curr_v = 0; // Current vertex

    while (!curr_path.empty())
    {
        // If there's remaining edge
        if (edge_count[curr_v])
        {
            // Push the vertex
            curr_path.push(curr_v);

            // Find the next vertex using an edge
            int next_v = l[curr_v].back();

            // and remove that edge
            edge_count[curr_v]--;
            l[curr_v].pop_back();

            // Move to next vertex
            curr_v = next_v;
        }

        // back-track to find remaining circuit
        else
        {
            circuit.push_back(curr_v);

            // Back-tracking
            curr_v = curr_path.top();
            curr_path.pop();
        }
    }

    return circuit;

    // we've got the circuit, now print it in reverse
    // for (int i = circuit.size() - 1; i >= 0; i--)
    // {
    //     cout << circuit[i];
    //     if (i)
    //         cout << " -> ";
    // }
}

int main()
{
    int n, fst, snd;
    vector<string> map;
    vector<vector<int>> l;
    vector<int> in;
    string temp;
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        cin >> temp;
        int j = 0;
        while (j < map.size() && temp.compare(1, 2, map[j]))
        {
            j++;
        }
        if (j == map.size())
        {
            map.push_back(temp.substr(1, 2));
        }
        snd = j;

        temp[2] = '\0';
        j = 0;
        while (j < map.size() && temp.compare(0, 2, map[j]))
        {
            j++;
        }
        if (j == map.size())
        {
            map.push_back(temp.substr(0, 2));
        }
        fst = j;

        while (fst >= l.size() || snd >= l.size())
        {
            l.push_back(vector<int>());
        }
        while (fst >= in.size() || snd >= in.size())
        {
            in.push_back(0);
        }
        
        l[fst].push_back(snd);
        in[snd]++;
    }
    /* for (auto &&i : l)
    {
        for (auto &&j : i)
        {
            cout << j << " ";
        }
        cout << "\n";
    } */
    
    int start = -1, end = -1;
    for (int i = 0; i < l.size(); i++)
    {
        if (l[i].size() == in[i] + 1)
        {
            if (start == -1)
            {
                start = i;
            }
            else
            {
                cout << "NO\n";
                return 0;
            }
        }
        else if (l[i].size() == in[i] - 1)
        {
            if (end == -1)
            {
                end = i;
            }
            else
            {
                cout << "NO\n";
                return 0;
            }
        }
        else if (l[i].size() != in[i])
        {
            cout << "NO\n";
            return 0;
        }
    }
    if (start != -1 && end != -1)
    {
        l[end].push_back(start);
    }
    
    // l represents the adjacency list of
    // the directed graph
    // edge_count represents the number of edges
    // emerging from a vertex
    vector<int> edge_count;

    for (int i = 0; i < l.size(); i++)
    {
        //find the count of edges to keep track
        //of unused edges
        edge_count.push_back(l[i].size());
    }

    // if (!adj.size())
    //     return; //empty graph

    // Maintain a stack to keep vertices
    stack<int> curr_path;

    // vector to store final circuit
    vector<int> circuit;

    // start from any vertex
    curr_path.push(0);
    int curr_v = 0; // Current vertex

    while (!curr_path.empty())
    {
        // If there's remaining edge
        if (edge_count[curr_v])
        {
            // Push the vertex
            curr_path.push(curr_v);

            // Find the next vertex using an edge
            int next_v = l[curr_v].back();

            // and remove that edge
            edge_count[curr_v]--;
            l[curr_v].pop_back();

            // Move to next vertex
            curr_v = next_v;
        }

        // back-track to find remaining circuit
        else
        {
            circuit.push_back(curr_v);

            // Back-tracking
            curr_v = curr_path.top();
            curr_path.pop();
        }
    }

    
    if (start == -1 && end == -1)
    {
        if (circuit.size() - 1 != n)
        {
            cout << "NO\n";
            return 0;
        }
        cout << "YES\n";
        for (int i = circuit.size()-1; i > 0; i--)
        {
            cout << map[circuit[i]][0];
        }
        cout << map[circuit[0]] << endl;
    }
    else
    {
        if (circuit.size() - 2 != n)
        {
            cout << "NO\n";
            return 0;
        }
        int i = circuit.size()-1;
        while (i > 0 && (circuit[i-1] != start || circuit[i] != end))
        {
            i--;
        }

        if (i == 0)
        {
            cout << "NO\n";
            return 0;
        }
        
        cout << "YES\n";
        for (int j = i-1; j > 0; j--)
        {
            cout << map[circuit[j]][0];
        }
        for (int j = circuit.size()-1; j > i; j--)
        {
            cout << map[circuit[j]][0];
        }
        cout << map[circuit[i]] << endl;
    }
    
    return 0;
}